### PR TITLE
[RLlib] Update bandit_envs_recommender_system

### PR DESCRIPTION
**Pull request by gjoliver**
## Why are these changes needed?

Update bandit_envs_recommender_system with Sven's in-house implementa…, which supports multiple users and slate sizes.
Make sure example tune_lin_ucb_train_recommendation.py nails this environment.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/examples/env/bandit_envs_recommender_system.py [hunk 1]
```diff
@@ -1,166 +1,227 @@
...
+            "response": response,
         }
-        if self.num_users > 1:
-            obs["user"] = self.current_user_id
-        return obs
 
-    def render(self, mode="human"):
-        raise NotImplementedError
+
+# Because ParametricRecSys follows RecSim's API, we have to wrap it before
```
Comment (sven1977) [inline rllib/examples/env/bandit_envs_recommender_system.py:194]: Should we do this here?
Maybe just move this into the block below and register it under "ParametricRecSysEnvForBandits" or something like this.
Also, in the example script that actually uses this env, we can register it to show, how this is done.

**Review by sven1977 — COMMENTED** [review]
Looks good, thanks for the cleanup @gjoliver ! :)  Left only some minor nits.

File: rllib/examples/env/bandit_envs_recommender_system.py (hunk 1 above)
Comment (gjoliver) [reply rllib/examples/env/bandit_envs_recommender_system.py]: I really like the idea of registering the env in example script. good demo.
thanks.

**Review by gjoliver — COMMENTED** [review]
done. thanks for the review.

**Comment by sven1977** [comment]
Hey @gjoliver , could you re-merge? Lots of tests failing, I'm thinking b/c we are too far behind master.

File: rllib/env/wrappers/recsim.py [hunk 2]
```diff
@@ -84,25 +84,20 @@ def __init__(self, env: gym.Env):
         self.observation_space = Dict(
             OrderedDict(
                 [
-                    ("user", obs_space["user"]),
                     (
                         "item",
                         gym.spaces.Box(
-                            low=-np.ones((num_items, embedding_dim)),
-                            high=np.ones((num_items, embedding_dim)),
+                            low=-1.0, high=1.0, shape=(num_items, embedding_dim)
```
Comment (sven1977) [inline rllib/env/wrappers/recsim.py:90]: Nice!

File: rllib/examples/env/bandit_envs_recommender_system.py [hunk 3]
```diff
@@ -1,166 +1,221 @@
...
+            self.current_user_budget -= 1.0
+        d = self.current_user_budget <= 0.0
+
+        # Compile response.
+        response = tuple(
+            {
+                "click": int(idx == which_clicked),
+                "engagement": r if idx == which_clicked else 0.0,
+            }
+            for idx in range(len(user_doc_overlaps) - 1)
```
Comment (sven1977) [inline rllib/examples/env/bandit_envs_recommender_system.py:159]:
```suggestion
        reward = 0.0
        if which_clicked < self.slate_size:
            # Reward is 1.0 - regret if clicked. 0.0 if not clicked.
            regret = best_reward - user_doc_overlaps[which_clicked]
            reward = 1 - regret
            # If anything clicked, deduct from the current user's time budget.
            self.current_user_budget -= 1.0
        done = self.current_user_budget <= 0.0

        # Compile response.
        response = tuple(
            {
                "click": int(idx == which_clicked),
                "engagement": reward if idx == which_clicked else 0.0,
            }
            for idx in range(len(user_doc_overlaps) - 1)
```

File: rllib/examples/env/bandit_envs_recommender_system.py [hunk 4]
```diff
@@ -1,166 +1,221 @@
...
-
-        if self.num_users == 1:
-            return spaces.Dict(
-                {
-                    "item": item_obs_space,
-                    "item_id": item_ids_obs_space,
-                    "response": resp_space,
-                }
-            )
+        return self._get_obs(response=response), r, d, {}
```
Comment (sven1977) [inline rllib/examples/env/bandit_envs_recommender_system.py:162]:
```suggestion
        return self._get_obs(response=response), reward, done, {}
```

**Review by sven1977 — APPROVED** [review]
Thanks for the fixes @gjoliver . Great PR! :)

**Pull request merged**